### PR TITLE
fix invalid memory access bug in `tidbGreatestString`

### LINKED ISSUE #8604: Memory control overestimates memory usage
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

table with 1000 rows join with table with 75 rows and join with table with 10 rows
Query failed for the the overestimated memory usage 4.00 EiB
```
explain analyze SELECT count(*) FROM table_1000_utf8mb4collate_utf8mb4_bin_undef t1 left join table_75_utf8mb4collate_utf8mb4_general_ci_undef t2 on t1. `col_timestamp_undef_signed` != t2. `col_int_key_unsigned` or not greatest(t1. `col_double_key_unsigned`,t1. `col_char(20)_key_signed`,null) inner join table_10_utf8mb4collate_utf8mb4_general_ci_undef t3 on t2. `col_int_undef_unsigned` != t3. `col_varchar(20)_key_signed`;
ERROR 1105 (HY000): other error for mpp stream: Code: 0, e.displayText() = DB::TiFlashException: Memory limit (total) exceeded caused by 'out of memory quota for data computing' : would use 4.00 EiB for data computing (attempt to allocate chunk of 4611686018427715620 bytes), limit of memory for data computing: 12.80 GiB. Memory Usage of Storage: non-query: peak=153.41 MiB, amount=371.88 KiB; kvstore: peak=196.10 KiB, amount=0.00 B; query-storage-task: peak=882.19 KiB, amount=0.00 B; fetch-pages: peak=0.00 B, amount=0.00 B
```
image

### 4. What is your TiFlash version? (Required)
master

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: close #8604

Problem Summary:
In `LeastGreatestStringImpl::mergeImpl`, it use memcpy to copy data to the target column, however, it does not reserve enough space before `LeastGreatestStringImpl::mergeImpl`, so there maybe some invalid memory read/write in this function, and will cause issue randomly.
### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Functions/LeastGreatest.h [hunk 1]
```diff
@@ -155,50 +154,37 @@ struct LeastGreatestStringImpl
...
             {
-                memcpy(&c_data[pre_offset], &a_data[0], a_size);
-                c_offsets[i] = pre_offset + a_size + 1;
-            }
-            else if (res == 0)
-            {
-                size_t size = std::min(a_size, b_size);
-                memcpy(&c_data[pre_offset], &b_data[0], size);
-                c_offsets[i] = pre_offset + size + 1;
+                append_data(&a_data[0], a_size);
```
Comment (gengliqi) [inline dbms/src/Functions/LeastGreatest.h:170]:
```suggestion
                append_data(a_data, a_size);
```

File: dbms/src/Functions/LeastGreatest.h [hunk 2]
```diff
@@ -155,50 +154,37 @@ struct LeastGreatestStringImpl
...
             }
             else
             {
-                memcpy(&c_data[pre_offset], &b_data[0], b_size);
-                c_offsets[i] = pre_offset + b_size + 1;
+                assert(res > 0 || a_size == b_size);
+                append_data(&b_data[0], b_size);
```
Comment (gengliqi) [inline dbms/src/Functions/LeastGreatest.h:175]:
```suggestion
                append_data(b_data, b_size);
```

File: dbms/src/Functions/LeastGreatest.h [hunk 3]
```diff
@@ -155,50 +154,37 @@ struct LeastGreatestStringImpl
...
-                {
-                    memcpy(&c_data[pre_offset], &a_data[0], size);
-                    c_offsets[i] = pre_offset + size + 1;
-                }
-                else
-                {
-                    memcpy(&c_data[pre_offset], &b_data[0], size);
-                    c_offsets[i] = pre_offset + size + 1;
-                }
+                append_data(&b_data[0], b_size);
```
Comment (gengliqi) [inline dbms/src/Functions/LeastGreatest.h:182]:
```suggestion
                append_data(b_data, b_size);
```

File: dbms/src/Functions/LeastGreatest.h [hunk 4]
```diff
@@ -155,50 +154,37 @@ struct LeastGreatestStringImpl
...
             }
             else
             {
-                memcpy(&c_data[pre_offset], &a_data[0], a_size);
-                c_offsets[i] = pre_offset + a_size + 1;
+                assert(res > 0 || a_size == b_size);
+                append_data(&a_data[0], a_size);
```
Comment (gengliqi) [inline dbms/src/Functions/LeastGreatest.h:187]:
```suggestion
                append_data(a_data, a_size);
```

**Review by gengliqi — APPROVED** [review]
Rest LGTM

**Review by yibin87 — APPROVED** [review]
LGTM

File: dbms/src/Functions/LeastGreatest.h (hunk 4 above)
Comment (windtalker) [reply dbms/src/Functions/LeastGreatest.h]: Done

**Pull request merged**